src/app/Rules/CreateRule.tsx

```diff
@@ -0,0 +1,374 @@
+  const [maxSizeUnits, setMaxSizeUnits] = React.useState(1);
+  const [archivalPeriod, setArchivalPeriod] = React.useState(0);
+  const [archivalPeriodUnits, setArchivalPeriodUnits] = React.useState(1);
+  const [preservedArchives, setPreservedArchives] = React.useState(0);
+
+  const handleNameChange = (evt) => {
+    setNameValid(RuleNamePattern.test(name) ? ValidatedOptions.success : ValidatedOptions.error);
+    setName(evt);
+  };
+
+  const getEventString = () => {
+    var str = '';
+    if (!!template) {
+      str += `template=${template}`;
+    }
+    if (!!templateType) {
+      str += `,type=${templateType}`;
+    }
+    return str;
+  };
+
+  const handleTemplateChange = (template) => {
+    const parts: string[] = template.split(',');
+    setTemplate(parts[0]);
+    setTemplateType(parts[1]);
+  };
+
+  const handleMaxAgeChange = (evt) => {
+    setMaxAge(Number(evt));
+  };
+
+  const handleMaxAgeUnitChange = (evt) => {
+    setMaxAgeUnits(Number(evt));
+  };
+
+  const handleMaxSizeChange = (evt) => {
+    setMaxSize(Number(evt));
+  };
+
+  const handleMaxSizeUnitChange = (evt) => {
+    setMaxSizeUnits(Number(evt));
+  };
+
+  const handleArchivalPeriodChange = (evt) => {
+    setArchivalPeriod(Number(evt));
+  };
+
+  const handleArchivalPeriodUnitsChange = (evt) => {
+    setArchivalPeriodUnits(Number(evt));
+  };
+
+  const handleSetPreservedArchives = (evt) => {
+    setPreservedArchives(Number(evt));
+  };
+
+  const handleSubmit = (): void => {
+    const notificationMessages: string[] = [];
+    if (nameValid !== ValidatedOptions.success) {
+      notificationMessages.push(`Rule name ${name} is invalid`);
+    }
+    if (notificationMessages.length > 0) {
+      const message = notificationMessages.join('. ').trim() + '.';
+      notifications.warning('Invalid form data', message);
+      return;
+    }
+    const rule: Rule = {
+      name,
+      description,
+      matchExpression,
+      eventSpecifier: getEventString(),
+      archivalPeriodSeconds: archivalPeriod * archivalPeriodUnits,
+      preservedArchives,
+      maxAgeSeconds: maxAge * maxAgeUnits,
+      maxSizeBytes: maxSize * maxSizeUnits
+    };
+    addSubscription(
+      context.api.createRule(rule)
+      .pipe(first())
+      .subscribe(success => {
+        if (success) {
+          history.push('/rules');
+        }
+      })
+    );
+  };
+
+  // FIXME we query ourselves to populate the list of templates, since Rules can apply to any target. Is this better than making the user write the event specifier manually,
+  // or at least make them write the name manually and choose TARGET/CUSTOM from a dropdown?
+  React.useEffect(() => {
+    addSubscription(
+      context.api.doGet<EventTemplate[]>(`targets/localhost:0/templates`).subscribe(setTemplates)
+    );
+  }, []);
+
+  const breadcrumbs: BreadcrumbTrail[] = [
+    {
+      title: 'Automated Rules',
+      path: '/rules',
+    }
+  ];
+
+  const templatesOptionGroups = React.useMemo(() => {
```

Comment:
Is this the same component used in the custom recording form? If so, could we extract and reuse it somehow?